vehicle/OVMS.V3/components/vehicle_nissanleaf/src/vehicle_nissanleaf.cpp

```diff
@@ -713,7 +721,17 @@ void OvmsVehicleNissanLeaf::IncomingFrameCan1(CAN_frame_t* p_frame)
           case 24: cd = CHARGE_DURATION_RANGE_L0; break;
           }
         if (cd != -1) m_charge_duration->SetElemValue(cd, val/2);
-        if (type != -1) m_battery_type->SetValue(type);
+        }
+      // If detected, save battery type
+      if (type != -1)
+        {
+        m_battery_type->SetValue(type);
+        }
+      else if (!m_battery_type->IsDefined())
```

Comment:
Not sure if the IsDefined() test still make sense.  It was previously needed to account for the additional information that might have turned up independently from the battery poll reply, but we don't have to worry about that now.
I think we can instead set type=BATTERY_TYPE_2_24kWh in switch(mx) cases 21 and 24, since TYPE_1_24kWh uses mx 10 and 18, and I believe your TYPE_2_30kWh never has valid values for either of these?

Reply:
Prefer to leave it as it is now. 
My Japanese TYPE2_30kwh has messages for mx cases 21 and 24, always with the invalid value 0xfff. I expect US models to behave the same. But I cannot yet rule out that UK 30kwh versions might show a valid value here. If so, then setting TYPE2_24kwh on these mx cases would be incorrect.

So for the moment I think it is safer to set a default of TYPE2_24kwh and only change it when we positively identify a TYPE1_24kwh or a TYPE2_30kwh.

Reply:
Did another google to figure out what the expected behaviour of 30kwh UK models would be.
Found 2 pages both indicating that the "long life mode (80%)" was removed from USA and Jap models somewhere in 2015 and from UK models as of mode year 2016.
    https://www.speakev.com/threads/long-life-mode-80-charge-no-longer-available.8261/
    https://samholford.github.io/leafguide/

So Robin is probably right and if we see a val!=0x1fff with mx=21 or 24 then this must be a type 2 battery from before 2016, which can only be a 24kwh.
Adapted the code accordingly and added comments describing this behavior.

Note that we do not have a way to positively identify a TYPE_2_24kwh from 2016 or 2017. So for the moment we still rely on using TYPE_2_24KWH as default when reading values from m_battery_type.